canvas_cli/tests.py

```diff
@@ -1,11 +1,170 @@
+            "Disable plugin",
+            f"disable {plugin_name}",
+            0,
+            [f"Disabling {plugin_name} using ", f"Plugin {plugin_name} successfully disabled!"],
+            [],
+        ),
+        ("List disabled plugin", "list", 0, [f"{plugin_name}@0.0.1	disabled"], []),
+        (
+            "Enable plugin",
+            f"enable {plugin_name}",
+            0,
+            [f"Enabling {plugin_name} using ", f"Plugin {plugin_name} successfully enabled!"],
+            [],
+        ),
+        (
+            "Uninstall enabled plugin",
+            f"uninstall {plugin_name}",
+            1,
+            [
+                f"Uninstalling {plugin_name} using",
+                'Status code 403: {"detail":"Cannot delete an enabled plugin."}',
+            ],
+            [],
+        ),
+        ("Disable plugin", f"disable {plugin_name}", 0, [], []),
+        (
+            "Uninstall disabled plugin",
+            f"uninstall {plugin_name}",
+            0,
+            [
+                f"Uninstalling {plugin_name} using",
+                f"Plugin {plugin_name} successfully uninstalled!",
+            ],
+            [],
+        ),
+        ("List empty plugins again", "list", 0, [], [f"{plugin_name}"]),
+    ]
+
+
+@patch("keyring.get_password")
+@patch("keyring.set_password")
+@pytest.mark.integtest
+def test_canvas_list_install_disable_enable_uninstall(
+    mock_get_password: MagicMock,
+    mock_set_password: MagicMock,
+    plugin_name: str,
+    cli_test_steps: list[tuple[str, str, int, list[str], list[str]]],
+) -> None:
+    """Tests that the Canvas CLI can list, install, disable, enable, and uninstall a plugin"""
+
+    mock_get_password.return_value = None
+    mock_set_password.return_value = None
+
+    runner.invoke(app, "init", input=plugin_name)
+    protocol = open(f"./{plugin_name}/protocols/my_protocol.py", "w")
+    p = [
+        "from canvas_sdk.events import EventType\n",
+        "from canvas_sdk.protocols import BaseProtocol\n",
+        "from logger import log\n",
+        "\n",
+        "\n",
+        "class Protocol(BaseProtocol):\n",
+        "\tRESPONDS_TO = EventType.Name(EventType.ASSESS_COMMAND__CONDITION_SELECTED)\n",
+        """\tNARRATIVE_STRING = "I was inserted from my plugin's protocol."\n""",
+        "\n",
+        "\tdef compute(self):\n",
+        "\t\tlog.info(self.NARRATIVE_STRING)\n",
+        "\t\treturn []\n",
+    ]
+    protocol.writelines(p)
+    protocol.close()
+
+    for (
+        step_name,
+        command,
+        expected_exit_code,
+        expected_outputs,
+        expected_no_outputs,
+    ) in cli_test_steps:
+        print(step_name)
+        result = runner.invoke(app, command)
+        assert result.exit_code == expected_exit_code
+        for expected_output in expected_outputs:
+            assert expected_output in result.stdout
+        for expected_no_ouput in expected_no_outputs:
+            assert expected_no_ouput not in result.stdout
```

Comment:
This might be better suited to test parametrization than a loop. If there is a failure it may not be easy to tell which one failed, but you can assign names to parametrization scenarios.

Reply:
i thought about that but they need to happen in order ... its a list of steps that happen one after the other. i think the parametrization all happens at the same time right?

Reply:
I believe they occur in order, but I see your point. Using parametrization for something that is more like setup wouldn't be appropriate. In that case, I'd probably suggest moving all of the invoking into a fixture, and then then parametrize all of the assertions.